feat(adapters): Add initial EmbeddingModel adapters

ajbozarth opened the pull request:
### Which issue(s) does this pull-request address?

Ref: #412

### Description

This includes the initial implementation of the `LiteLLMEmbeddingModel`, `OllamaEmbeddingModel` and `WatsonxEmbeddingModel` adapters.

It also reintroduces `Backend` which was removed from #787 prior to merge 

### Checklist

#### General
- [x] I have read the appropriate contributor guide: [Python](https://github.com/i-am-bee/beeai-framework/blob/main/CONTRIBUTING.md)
/ [TypeScript](https://github.com/i-am-bee/beeai-framework/blob/main/CONTRIBUTING.md)
- [x] I have signed off on my commit: [Python instructions](https://github.com/i-am-bee/beeai-framework/blob/main/python/CONTRIBUTING.md#developer-certificate-of-origin-dco) / [TypeScript instructions](https://github.com/i-am-bee/beeai-framework/blob/main/typescript/CONTRIBUTING.md#developer-certificate-of-origin-dco)
- [x] Commit messages and PR title follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- [x] Appropriate label(s) added to PR: `Python` for Python changes, `TypeScript` for TypeScript changes

#### Code quality checks
- [x] Linting passes: Python `poe lint` or `poe lint --fix` / TypeScript `yarn lint` or `yarn lint:fix`
- [x] Formatting is applied: Python `poe format` or `poe format --fix` / TypeScript: `yarn format` or `yarn format:fix`
- [x] Static type checks pass: Python `poe type-check`

#### Testing
- [x] Unit tests pass: Python `poe test --type unit` / TypeScript `yarn test:unit`
- [x] E2E tests pass: Python `poe test --type e2e` / TypeScript: `yarn test:e2e`
- [x] Integration tests pass: Python `poe test --type integration`
- [x] Tests are included (for bug fixes or new features)

#### Documentation
- [x] Documentation is updated
- [x] Embedme embeds code examples in docs. To update after edits, run: Python `poe docs --type build`

ajbozarth commented:
This is my initial attempt at an EmbeddingModel adapter. I assume it needs polishing, if not significant improvements. But it is functional as is for Ollama.

My next steps are to add the `WatsonxEmbeddingModel` and update the docs

Tomas2D commented on python/beeai_framework/adapters/litellm/embedding.py:
```diff
@@ -0,0 +1,134 @@
+
+    def _transform_input(self, model_input: EmbeddingModelInput) -> dict[str, Any]:
+        return {
+            "model": self.provider_id + "/" + self._model_id,
```
use f-string

Tomas2D commented on python/beeai_framework/adapters/litellm/embedding.py:
```diff
@@ -0,0 +1,134 @@
+
+    @staticmethod
+    def litellm_debug(enable: bool = True) -> None:
+        litellm.set_verbose = enable  # type: ignore [attr-defined]
+        litellm.suppress_debug_info = not enable
+
+        litellm.suppress_debug_info = not enable
+        litellm.logging = enable
+
+        logger = logging.getLogger("LiteLLM")
+        logger.setLevel(logging.DEBUG if enable else logging.CRITICAL + 1)
```
This is duplicated. Extract this to a function shared between `LiteLLMChatModel` and `LiteLLMEmbeddingModel`. It can be stored in `adapters/litellm/utils.py`/

Tomas2D commented on python/beeai_framework/adapters/litellm/embedding.py:
```diff
@@ -0,0 +1,134 @@
+        embeddings: list[list[float]] = []
+
+        for embedding in response.data:
+            embeddings.append([float(point) for point in embedding.get("embedding")])
```
```python
for result in response.data:
    embeddings.append([float(value) for value in result["embedding"]])
```

Tomas2D commented on python/beeai_framework/adapters/litellm/embedding.py:
```diff
@@ -0,0 +1,134 @@
+        return EmbeddingModelOutput(
+            values=model_input.values,
+            embeddings=embeddings,
+            usage=EmbeddingModelUsage(**response.usage.model_dump()) if response.usage else None,
```
I would extend the `EmbeddingModelOutput` to include a raw response (untyped obviously).

ajbozarth commented on python/beeai_framework/adapters/litellm/utils.py:
```diff
@@ -56,6 +56,10 @@ def parse_extra_headers(
 
 def _parse_header_string(header_string: str) -> dict[str, str]:
     """Parses a comma-separated string of headers into a dictionary."""
+    from beeai_framework.logger import Logger
```
I moved this inside the function because the new calls to `litellm_debug` in this file were causing a import loop for `logger`

ajbozarth commented:
I've added the `WatsonxEmbeddingModel` and addressed the current review. That adds the adapters specifically requested in #412 and the only adapters I have access to an embedding model instance for (the service I used to test OpenAPI doesn't have embedding models)

Tomas2D reviewed (approved):
It works. Thank you!
